`inline given ... with` not compiling properly

## Compiler version

3.0.0, 3.0.1, 3.1.0, 3.1.1-RC2

## Minimized code

```Scala
import scala.compiletime.*

trait C[A]

inline given [Tup <: Tuple]: C[Tup] with
  val cs = summonAll[Tuple.Map[Tup, C]]

```

## Output

Compilation error:

```scala
method summonAll is declared as `inline`, but was not inlined
```

## Expectation

Compile successfully, or at least provide a better error message if the construct is indeed invalid.

When a `.toList` is added to the last line, the compiler crashes:

```scala
inline given [Tup <: Tuple]: C[Tup] with
  val cs = summonAll[Tuple.Map[Tup, C]].toList
```

```
exception while typing scala.compiletime.summonAll[Tuple.Map[Tup, bug.C]].toList of class class dotty.tools.dotc.ast.Trees$Select # -1
exception while typing def cs:

    scala.collection.immutable.List[
      Tuple.Map[Tup, bug.C] match {
        case EmptyTuple => Nothing
        case h *: t => h | scala.Tuple.Fold[t, Nothing, [x, y] =>> x | y]
      }
    ]

 = scala.compiletime.summonAll[Tuple.Map[Tup, bug.C]].toList of class class dotty.tools.dotc.ast.Trees$DefDef # -1
exception while typing given class given_C_Tup[Tup <: Tuple]() extends Object(), bug.C[Tup] {
  private type Tup <: Tuple
  def cs:

      scala.collection.immutable.List[
        Tuple.Map[Tup, bug.C] match {
          case EmptyTuple => Nothing
          case h *: t => h | scala.Tuple.Fold[t, Nothing, [x, y] =>> x | y]
        }
      ]

   = scala.compiletime.summonAll[Tuple.Map[Tup, bug.C]].toList
} of class class dotty.tools.dotc.ast.Trees$TypeDef # -1
exception while typing @SourceFile("src/main/scala/bug.scala") final module class bug$package() extends

Object() {
  private def writeReplace(): AnyRef =
    new scala.runtime.ModuleSerializationProxy(classOf[bug.bug$package.type])
  given class given_C_Tup[Tup <: Tuple]() extends Object(), bug.C[Tup] {
    private type Tup <: Tuple
    def cs:

        scala.collection.immutable.List[
          Tuple.Map[Tup, bug.C] match {
            case EmptyTuple => Nothing
            case h *: t => h | scala.Tuple.Fold[t, Nothing, [x, y] =>> x | y]
          }
        ]

     = scala.compiletime.summonAll[Tuple.Map[Tup, bug.C]].toList
  }
  final given def given_C_Tup[Tup <: Tuple]: bug.bug$package.given_C_Tup[Tup] =
    new ([Tup <: Tuple] =>> bug.bug$package.given_C_Tup[Tup])[Tup]()
} of class class dotty.tools.dotc.ast.Trees$TypeDef # -1
exception while typing package bug {
  @SourceFile("src/main/scala/bug.scala") trait C[A]() extends Object {
    private type A
  }
  final lazy module val bug$package: bug.bug$package = new bug.bug$package()
  @SourceFile("src/main/scala/bug.scala") final module class bug$package()
     extends
   Object() {
    private def writeReplace(): AnyRef =
      new scala.runtime.ModuleSerializationProxy(classOf[bug.bug$package.type])
    given class given_C_Tup[Tup <: Tuple]() extends Object(), bug.C[Tup] {
      private type Tup <: Tuple
      def cs:

          scala.collection.immutable.List[
            Tuple.Map[Tup, bug.C] match {
              case EmptyTuple => Nothing
              case h *: t => h | scala.Tuple.Fold[t, Nothing, [x, y] =>> x | y]
            }
          ]

       = scala.compiletime.summonAll[Tuple.Map[Tup, bug.C]].toList
    }
    final given def given_C_Tup[Tup <: Tuple]: bug.bug$package.given_C_Tup[Tup]
       =
    new ([Tup <: Tuple] =>> bug.bug$package.given_C_Tup[Tup])[Tup]()
  }
} of class class dotty.tools.dotc.ast.Trees$PackageDef # -1
[info] exception occurred while compiling E:\prog\scala\typelevel\src\main\scala\bug.scala
java.lang.AssertionError: assertion failed: method toList in trait Tuple while compiling E:\prog\scala\typelevel\src\main\scala\bug.scala
[error] ## Exception when compiling 2 sources to E:\prog\scala\typelevel\target\scala-3.1.1-RC2\classes
[error] java.lang.AssertionError: assertion failed: method toList in trait Tuple
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] dotty.tools.dotc.transform.Erasure$Typer.mapOwner$1(Erasure.scala:714)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedSelect(Erasure.scala:726)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2749)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2841)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:121)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2907)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2911)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3027)
[error] dotty.tools.dotc.typer.Typer.$anonfun$46(Typer.scala:2228)
[error] dotty.tools.dotc.typer.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:238)
[error] dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:2228)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedDefDef(Erasure.scala:962)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2756)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2841)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:121)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2907)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2911)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2933)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2983)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1060)
[error] dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2426)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedClassDef(Erasure.scala:1050)
[error] dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:2767)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2771)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2841)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:121)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2907)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2911)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2933)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2983)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1060)
[error] dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2426)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedClassDef(Erasure.scala:1050)
[error] dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:2767)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2771)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2841)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:121)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2907)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2911)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2933)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2983)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1060)
[error] dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2553)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2812)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2842)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:121)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2907)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2911)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3027)
[error] dotty.tools.dotc.transform.Erasure.run(Erasure.scala:132)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:308)
[error] scala.collection.immutable.List.map(List.scala:246)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:309)
[error] dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:261)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$1(Run.scala:272)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:280)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:289)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:222)
[error] dotty.tools.dotc.Run.compile(Run.scala:206)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:192)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:247)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:182)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:210)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:177)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:175)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:461)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:416)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:503)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:403)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:169)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:482)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2366)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2316)
[error] sbt.internal.server.BspCompileTask$.$anonfun$compute$1(BspCompileTask.scala:30)
[error] sbt.internal.io.Retry$.apply(Retry.scala:46)
[error] sbt.internal.io.Retry$.apply(Retry.scala:28)
[error] sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:30)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2314)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] sbt.Execute.work(Execute.scala:291)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] java.lang.Thread.run(Thread.java:748)
```

--- a/compiler/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/compiler/src/dotty/tools/dotc/ast/Desugar.scala
@@ -809,7 +809,9 @@ object desugar {
             Nil
         }
       }
-      val classMods = if mods.is(Given) then mods &~ Inline | Synthetic else mods
+      if mods.isAllOf(Given | Inline | Transparent) then
+        report.error("inline non-alias given cannot be trasparent", cdef)
+      val classMods = if mods.is(Given) then mods &~ (Inline | Transparent) | Synthetic else mods
       cpy.TypeDef(cdef: TypeDef)(
         name = className,
         rhs = cpy.Template(impl)(constr, parents1, clsDerived, self1,

--- a/compiler/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/compiler/src/dotty/tools/dotc/ast/Desugar.scala
@@ -783,7 +783,7 @@ object desugar {
         DefDef(
           className.toTermName, joinParams(constrTparams, defParamss),
           classTypeRef, creatorExpr)
-          .withMods(companionMods | mods.flags.toTermFlags & GivenOrImplicit | Final)
+          .withMods(companionMods | mods.flags.toTermFlags & (GivenOrImplicit | Inline) | Final)
           .withSpan(cdef.span) :: Nil
       }
 
@@ -809,7 +809,7 @@ object desugar {
             Nil
         }
       }
-      val classMods = if mods.is(Given) then mods | Synthetic else mods
+      val classMods = if mods.is(Given) then mods &~ Inline | Synthetic else mods
       cpy.TypeDef(cdef: TypeDef)(
         name = className,
         rhs = cpy.Template(impl)(constr, parents1, clsDerived, self1,

--- a/compiler/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/compiler/src/dotty/tools/dotc/ast/Desugar.scala
@@ -810,7 +810,7 @@ object desugar {
         }
       }
       if mods.isAllOf(Given | Inline | Transparent) then
-        report.error("inline non-alias given cannot be trasparent", cdef)
+        report.error("inline given instances cannot be trasparent", cdef)
       val classMods = if mods.is(Given) then mods &~ (Inline | Transparent) | Synthetic else mods
       cpy.TypeDef(cdef: TypeDef)(
         name = className,

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -3626,7 +3626,7 @@ object Parsers {
           val templ =
             if isStatSep || isStatSeqEnd then Template(constr, parents, Nil, EmptyValDef, Nil)
             else withTemplate(constr, parents)
-          if noParams then ModuleDef(name, templ)
+          if noParams && !mods.is(Inline) then ModuleDef(name, templ)
           else TypeDef(name.toTypeName, templ)
       end gdef
       finalizeDef(gdef, mods1, start)

--- a/docs/docs/reference/contextual/givens.md
+++ b/docs/docs/reference/contextual/givens.md
@@ -114,6 +114,9 @@ inline given Show[Foo] with {
 ```
 Note that the inline methods within the given instances may be `transparent`.
 
+The inlining of given instances will not inline/duplicate the implementation of the given, it will just inline the instantiation of that instance.
+This is used to help dead code elimination of the given instances that are not used after inlining.
+
 
 ## Pattern-Bound Given Instances
 

--- a/docs/docs/reference/contextual/givens.md
+++ b/docs/docs/reference/contextual/givens.md
@@ -100,6 +100,21 @@ transparent inline given mkAnnotations[A, T]: Annotations[A, T] = ${
 
 Since `mkAnnotations` is `transparent`, the type of an application is the type of its right-hand side, which can be a proper subtype of the declared result type `Annotations[A, T]`.
 
+Given instances can have the `inline` but not `transparent` modifiers as their type is already known from the signature.
+Example:
+
+```scala
+trait Show[T] {
+  inline def show(x: T): String
+}
+
+inline given Show[Foo] with {
+  /*transparent*/ inline def show(x: Foo): String = ${ ... }
+}
+```
+Note that the inline methods within the given instances may be `transparent`.
+
+
 ## Pattern-Bound Given Instances
 
 Given instances can also appear in patterns. Example:

--- a/docs/docs/reference/contextual/givens.md
+++ b/docs/docs/reference/contextual/givens.md
@@ -111,6 +111,10 @@ trait Show[T] {
 inline given Show[Foo] with {
   /*transparent*/ inline def show(x: Foo): String = ${ ... }
 }
+
+def app =
+  // inlines `show` method call and removes the call to `given Show[Foo]`
+  summon[Show[Foo]].show(foo)
 ```
 Note that the inline methods within the given instances may be `transparent`.
 

--- a/tests/neg/i14177a.scala
+++ b/tests/neg/i14177a.scala
@@ -1,0 +1,6 @@
+import scala.compiletime.*
+
+trait C[A]
+
+inline given [Tup <: Tuple]: C[Tup] with
+  val cs = summonAll[Tuple.Map[Tup, C]] // error cannot reduce inline match with

--- a/tests/neg/i14177c.scala
+++ b/tests/neg/i14177c.scala
@@ -1,0 +1,15 @@
+class T
+
+transparent inline given fail1: T with // error
+  val cs = scala.compiletime.summonAll[EmptyTuple]
+transparent inline given fail2[X]: T with // error
+  val cs = scala.compiletime.summonAll[EmptyTuple]
+transparent inline given fail3(using DummyImplicit): T with // error
+  val cs = scala.compiletime.summonAll[EmptyTuple]
+
+transparent inline given ok1: T = new T:
+  val cs = scala.compiletime.summonAll[EmptyTuple]
+transparent inline given ok2[X]: T = new T:
+  val cs = scala.compiletime.summonAll[EmptyTuple]
+transparent inline given ok3(using DummyImplicit): T = new T:
+  val cs = scala.compiletime.summonAll[EmptyTuple]

--- a/tests/pos/i14177b.scala
+++ b/tests/pos/i14177b.scala
@@ -1,0 +1,15 @@
+class T
+
+inline given fail1: T with
+  val cs = scala.compiletime.summonAll[EmptyTuple]
+inline given fail2[X]: T with
+  val cs = scala.compiletime.summonAll[EmptyTuple]
+inline given fail3(using DummyImplicit): T with
+  val cs = scala.compiletime.summonAll[EmptyTuple]
+
+inline given ok1: T = new T:
+  val cs = scala.compiletime.summonAll[EmptyTuple]
+inline given ok2[X]: T = new T:
+  val cs = scala.compiletime.summonAll[EmptyTuple]
+inline given ok3(using DummyImplicit): T = new T:
+  val cs = scala.compiletime.summonAll[EmptyTuple]

--- a/tests/pos/i14282.scala
+++ b/tests/pos/i14282.scala
@@ -1,0 +1,13 @@
+trait Foo[A] {
+  inline def foo(): Unit
+}
+
+inline given FooA[A]: Foo[A] with {
+  inline def foo(): Unit = println()
+}
+def test1 = FooA.foo()
+
+inline given FooInt: Foo[Int] with {
+  inline def foo(): Unit = println()
+}
+def test2 = FooInt.foo()

